### PR TITLE
add adjoint, hat and vee for SE3

### PR DESCRIPTION
**Pull request by fantaosha**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: theseus/geometry/se3.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+from .lie_group import LieGroup
+from .point_types import Point3
+from .so3 import SO3
+
+
+class SE3(LieGroup):
+    def __init__(
+        self,
+        x_y_z_quaternion: Optional[torch.Tensor] = None,
```
Comment (mhmukadam) [inline theseus/geometry/se3.py:21]: Given that we follow _rotation-then-translation_ convention (like gtsam), should we make this input consistent with the convention.

File: theseus/geometry/se3.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        data: Optional[torch.Tensor] = None,
+        name: Optional[str] = None,
+        dtype: Optional[torch.dtype] = None,
+    ):
+        if x_y_z_quaternion is not None and data is not None:
+            raise ValueError("Please provide only one of x_y_z_quaternion or data.")
```
Comment (mhmukadam) [inline theseus/geometry/se3.py:27]: Should we be more explicit here and for other Lie groups what we mean by data? For instance here it is a `bx4x4` transformation matrix.

File: theseus/geometry/se3.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        return SE3(data=ret)
+
+    def to_matrix(self) -> torch.Tensor:
+        ret = torch.zeros(self.shape[0], 4, 4).to(dtype=self.dtype, device=self.device)
+        ret[:, :3] = self.data
+        ret[:, 3, 3] = 1
+        return ret
+
+    def update_from_x_y_z_quaternion(self, x_y_z_quaternion: torch.Tensor):
+        if x_y_z_quaternion.ndim != 2 and x_y_z_quaternion.shape[1] != 7:
```
Comment (mhmukadam) [inline theseus/geometry/se3.py:105]: We want to handle the case where the batch dimension is not passed and therefore batch size is assumed to be 1 (possibly also check in other places if we account for this).

File: theseus/geometry/se3.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        batch_size = x_y_z_quaternion.shape[0]
+        self.data = torch.empty(batch_size, 3, 4).to(
+            device=x_y_z_quaternion.device, dtype=x_y_z_quaternion.dtype
+        )
+        self[:, :3, :3] = SO3.unit_quaternion_to_matrix(x_y_z_quaternion[:, 3:])
+        self[:, :3, 3] = x_y_z_quaternion[:, :3]
+
+    def update_from_rot_and_trans(self, rotation: SO3, translation: Point3):
+        if rotation.shape[0] != translation.shape[0]:
+            raise ValueError("rotation and translation must have the same size.")
```
Comment (mhmukadam) [inline theseus/geometry/se3.py:117]: same size -> same batch size

File: theseus/geometry/se3.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        return cast(SE3, super().copy(new_name=new_name))
+
+    def transform_to(
+        self,
+        point: Union[Point3, torch.Tensor],
+        jacobians: Optional[List[torch.Tensor]] = None,
+    ) -> Point3:
+        raise NotImplementedError
+
+    def untransform_to(
```
Comment (mhmukadam) [inline theseus/geometry/se3.py:162]: Do you mean `transform_from` or is this different?

**Review by mhmukadam — APPROVED** [review]
Looks good!

File: theseus/geometry/se3.py (hunk 1 above)
Comment (fantaosha) [reply theseus/geometry/se3.py:21]: This is consistent with the convention in [`SE2`](https://github.com/facebookresearch/theseus/blob/6a92aff8a924e1bf036118d305a5afb3ff195a86/theseus/geometry/se2.py#L24), which uses `x_y_theta` for initialization. I'm OK with either.

Let me know if you would like to change the convention.

File: theseus/geometry/se3.py (hunk 2 above)
Comment (fantaosha) [reply theseus/geometry/se3.py:27]: This line is only used to resolve the conflict where both `x_y_z_quaternion` and `data` are provided.

There is another [function](https://github.com/facebookresearch/theseus/blob/f8cf872dc3d3acffc3232eace5826c6ef486babb/theseus/geometry/se3.py#L62) that explicitly defines what `data` should look like.

File: theseus/geometry/se3.py (hunk 3 above)
Comment (fantaosha) [reply theseus/geometry/se3.py:105]: OK. I will add this feature.

File: theseus/geometry/se3.py (hunk 5 above)
Comment (fantaosha) [reply theseus/geometry/se3.py:162]: Yeah, I mean `transform_from` and will change the name. 

Also, it seems that no `transform_from` is defined in [`SE2`](https://github.com/facebookresearch/theseus/blob/taoshaf.add_se3/theseus/geometry/se2.py). Should we add it to `SE2`.

File: theseus/geometry/se3.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    ):
+        if x_y_z_quaternion is not None and data is not None:
+            raise ValueError("Please provide only one of x_y_z_quaternion or data.")
+        if x_y_z_quaternion is not None:
+            dtype = x_y_z_quaternion.dtype
+        if data is not None:
+            self._SE3_matrix_check(data)
+        super().__init__(data=data, name=name, dtype=dtype)
+        if x_y_z_quaternion is not None:
+            if x_y_z_quaternion is not None:
```
Comment (luisenp) [inline theseus/geometry/se3.py:34]: Repeated if statement.

File: theseus/geometry/se3.py (hunk 3 above)
Comment (luisenp) [reply theseus/geometry/se3.py:105]: TBH, I'm not sure if we support this in 2D functions. Should we open an issue to remember to check for this?

File: theseus/geometry/se3.py (hunk 5 above)
Comment (luisenp) [reply theseus/geometry/se3.py:162]: Created issue #69 for this to address in a separate PR. I'd say this is lower priority, so it can wait until the 3D groups are finished.

**Review by luisenp — APPROVED** [review]
LGTM!

File: theseus/geometry/se3.py (hunk 4 above)
Comment (fantaosha) [reply theseus/geometry/se3.py:117]: Fixed.

File: theseus/geometry/se3.py (hunk 6 above)
Comment (fantaosha) [reply theseus/geometry/se3.py:34]: Fixed.

File: theseus/geometry/se3.py (hunk 3 above)
Comment (fantaosha) [reply theseus/geometry/se3.py]: There is `SO2.update_from_angle` but `SE2` seems to have no similar functions.

File: theseus/geometry/se3.py (hunk 1 above)
Comment (mhmukadam) [reply theseus/geometry/se3.py:21]: SE3 data is rotation-then-translation. @luisenp, @dongjing3309 any thoughts?

File: theseus/geometry/se3.py (hunk 2 above)
Comment (mhmukadam) [reply theseus/geometry/se3.py:27]: That makes sense. I meant, when this error statement is printed the user will have an opaque prompt for `data`.

File: theseus/geometry/se3.py (hunk 3 above)
Comment (mhmukadam) [reply theseus/geometry/se3.py]: > TBH, I'm not sure if we support this in 2D functions. Should we open an issue to remember to check for this?

Okay with either of these but we should be consistent:
a. We clarify in docs that batch dimension is always present and users need to add a stub dim for a batch size of one.
b. We let users define things without batch and append the extra dim internally everywhere.

File: theseus/geometry/se3.py [hunk 7]
```diff
@@ -123,7 +122,7 @@ def update_from_x_y_z_quaternion(self, x_y_z_quaternion: torch.Tensor):
 
     def update_from_rot_and_trans(self, rotation: SO3, translation: Point3):
         if rotation.shape[0] != translation.shape[0]:
-            raise ValueError("rotation and translation must have the same size.")
+            raise ValueError("rotation and translation must have the same batch size.")
```
Comment (mhmukadam) [inline theseus/geometry/se3.py:125]: Flagging this for the scenario when a user might pass rotation and translation without a batch dim.

File: theseus/geometry/se3.py (hunk 3 above)
Comment (luisenp) [reply theseus/geometry/se3.py]: Actually, for some reason I was thinking of something else. In Lie group classes constructors we do automatically append batch dimensions if not present, and @fantaosha also added this here in [line 35](https://github.com/facebookresearch/theseus/pull/68/files/f8cf872dc3d3acffc3232eace5826c6ef486babb#diff-c5e3f8d4504d1054716f8e6877b315b545ff52580232d8433d837c74e0ef4461R35). I guess the question is if it makes sense to move the `unsqueeze` inside the `updated_from_x_y_z_quaternion` function, which I think it does.

File: theseus/geometry/se3.py (hunk 1 above)
Comment (luisenp) [reply theseus/geometry/se3.py:21]: I don't have a strong preference here. Aesthetically it seems more pleasing to me to be consistent between `SE2` and `SE3` (even if it means refactoring `SE2` to use `theta_x_y`). But we should def give priority to any well-established conventions for working with se2 and se3 objects over this.

File: theseus/geometry/se3.py (hunk 2 above)
Comment (luisenp) [reply theseus/geometry/se3.py:27]: I think this is fine as is. The format for data will be documented and there are also other exceptions thrown if the user provides the wrong format for data.

File: theseus/geometry/se3.py (hunk 1 above)
Comment (luisenp) [reply theseus/geometry/se3.py:21]: Created #76 for this, since it also involves SE2 changes.

File: theseus/geometry/se3.py (hunk 3 above)
Comment (luisenp) [reply theseus/geometry/se3.py]: Let's move the `unsqueeze` for adding a batch dimension (if needed) inside `updated_from_x_y_z_quaternion` before merging.

File: theseus/geometry/se3.py (hunk 7 above)
Comment (luisenp) [reply theseus/geometry/se3.py:125]: Since inputs are `SO3` and `Point3`, they would already have a batch dim here, but we should add some type checks.

File: theseus/geometry/se3.py (hunk 3 above)
Comment (mhmukadam) [reply theseus/geometry/se3.py]: @fantaosha @luisenp was this addressed before merging? If not maybe add to an active PR.

File: theseus/geometry/se3.py (hunk 7 above)
Comment (mhmukadam) [reply theseus/geometry/se3.py:125]: @fantaosha @luisenp was this addressed before merging? If not, maybe add to an active PR.

Comment (luisenp) [reply theseus/geometry/se3.py:125]: Yes, I meant a type check. But I don't think it's super important to have this, it'd be mostly for the user to get a less confusing error message, but I don't expect this function to be used a lot directly.
